### PR TITLE
BUGFIX 4298 content stream creation deduplication

### LINKED ISSUE #4298: BUG: Importing a dump via `cr:import` will create duplicate content stream entries
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Using `cr:import` instead of `cr:migratelegacydata` to set up a neos 9 instance will result in:
- there will be **two** recorded events of type `ContentStreamWasCreated` in `cr_default_events`
- the table `cr_default_p_contentstream` will have two entries with the same `contentStreamId` and different `states`

Example import result:
| contentStreamId | version | sourceContentStreamId | state | removed |
| :--- | :--- | :--- | :--- | :--- |
| 80efbe37-9642-4109-8448-c09db47523aa | 14 | null | IN\_USE\_BY\_WORKSPACE | 0 |
| 80efbe37-9642-4109-8448-c09db47523aa | 14 | null | CREATED | 0 |

_Why does this matter?_

I stumbled upon this while using the cli command `./flow contentstream:prune --remove-temporary` based on the data dump included in `Neos.Demo`. Running `./flow contentstream:prune --remove-temporary`  removed the live workspace in my test installation so I investigated what caused this.

The `ContentStreamFinder` will remove the temporary entries of `state` `CREATED` or `FORKED` so why are there multiple entries for the live workspace?

_Why does this happen?_

The `cr:export` command will export all events from the contentstream connected to the `live` workspace. This includes the first event `ContentStreamWasCreated`, which in comparison is not included in the events generated by `cr:migratelegacydata`.

As the `EventStoreImportProcessor` creates the contentstream and the live workspace before applying all events to import we end up with the described result.

https://github.com/neos/neos-development-collection/blob/26adf216174f4446c0c46483e8d7fa51dad8c15c/Neos.ContentRepository.Export/src/Processors/EventStoreImportProcessor.php#L113-L143

### Expected Behavior

The contentstream for the live workspace should only be created once.

### Steps To Reproduce

**Reproduce with a newlyy created site:** 

1. Have an empty neos 9 project with the Neos.Demo package at hand
2. Setup the CR (`./flow cr:setup`)
3. Create a new site via `./flow site:create neosdemo Neos.Demo Neos.Demo:Document.Homepage`
4. Export the CR via `./flow cr:export --path=./crExportTest`
5. Cleanup your CR for reimporting the export via `./flow cr:prune`
6. Import via `./flow cr:import --path=./crExportTest`

**Reproduce via Neos.Demo dump:**

1. Have an empty neos 9 project with the Neos.Demo package at hand
2. Setup the CR is set up (`./flow cr:setup`)
3. Make sure there is an active `Site` in your installation (`./flow site:create neosdemo Neos.Demo Neos.Demo:Document.Homepage`)
4. Empty your CR via `./flow cr:prune`
5. Import all content from the Neos.Demo site package via `./flow cr:import ./Packages/Sites/Neos.Demo/Resources/Private/Content/`

### Environment

```markdown
- Flow: 9.0
- Neos: 9.0
- PHP: 8.2
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nezaniel**
Resolves #4298 

This removes the initial content stream creation event from the import process, as it is manually added anyway

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 branch
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Review by ahaeslich — APPROVED** [review]
Tested with both commands `cr:import` and `cr:migratelegacydata`. Thx for taking care.

File: Neos.Neos/Tests/Functional/Command/CrCommandControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        $this->contentRepositoryRegistry->get(ContentRepositoryId::fromString('default'))->setUp();
+        $sourcePath = __DIR__ . '/../Fixtures';
+        $this->subject->importCommand($sourcePath);
+        $targetPath = FLOW_PATH_DATA . 'Temporary';
+        $this->subject->exportCommand($targetPath);
+
+        $expectedEvents = Arrays::trimExplode("\n", file_get_contents($sourcePath. '/events.jsonl'));
+        $actualEvents = Arrays::trimExplode("\n", file_get_contents($targetPath. '/events.jsonl'));
+
+        self::assertSame(count($expectedEvents), count($actualEvents));
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Functional/Command/CrCommandControllerTest.php:47]: id like some stricter comparison here... not just compare the number of lines.

File: Neos.Neos/Tests/Behavior/Features/FrontendRouting/DisableNodes.feature [hunk 2]
```diff
@@ -158,37 +158,6 @@ Feature: Routing behavior of removed, disabled and re-enabled nodes
...
-      | Key                          | Value                    |
-      | contentStreamId              | "cs-identifier"          |
-      | nodeAggregateId              | "sir-david-nodenborough" |
-      | coveredDimensionSpacePoint   | {}                       |
-      | nodeVariantSelectionStrategy | "allVariants"            |
-    And the event NodeAggregateWasDisabled was published with payload:
-      | Key                          | Value                    |
-      | contentStreamId              | "cs-identifier"          |
-      | nodeAggregateId              | "sir-david-nodenborough" |
-      | affectedDimensionSpacePoints | [{}]                     |
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/FrontendRouting/DisableNodes.feature:172]: so this state should be impossible, is thats why this test was removed?

Comment (nezaniel) [reply Neos.Neos/Tests/Behavior/Features/FrontendRouting/DisableNodes.feature:172]: yep, this is to be caught by constraint checks

**Comment by mhsdesign** [comment]
Hmmm

> The cr:export command will export all events from the contentstream connected to the live workspace. This includes the first event ContentStreamWasCreated, which in comparison is not included in the events generated by cr:migratelegacydata.

but why dont we fix this and include the contentstream creation in both files? Wouldnt that reduce the conditional complexity here?

**Comment by kitsunet** [comment]
That seems logical, I guess we can fake that event, can't we?

**Comment by ahaeslich** [comment]
> The cr:export command will export all events from the contentstream connected to the live workspace. This includes the first event ContentStreamWasCreated, which in comparison is not included in the events generated by cr:migratelegacydata.

Taking this out of context might be missleading. This was just an observation why this behaviour might have been missed. The real issue is that the import based on an export already containing an event `ContentStreamWasCreated` for the live workspace will create a dublicate entry in `cr_default_p_contentstream` (see example in issue). 

Therefore adding the same event to `cr:migratelegacydata` won't fix the [issue](https://github.com/neos/neos-development-collection/issues/4298). 

**The TL;DR summary of the bug is: we must avoid dublication of the event creating the live contentstream.**

**Comment by mhsdesign** [comment]
So @ahaeslich if we remove the first `ContentStreamWasCreated` event from the export in the Neos.Demo

https://github.com/neos/Neos.Demo/blob/1c38bc76eeac37bebb8e09074e71a9893b1ee245/Resources/Private/Content/events.jsonl#L1

and take care of the exporter not emitting this even we would also be on the safe side i presume.

I would prefer that solution over ignoring events on import as this becomes harder to understand.

**Comment by mhsdesign** [comment]
Ping i think i need this as i adjusted the demo export and due to the re im and export there will be now 3 creation events 😂 
https://github.com/neos/Neos.Demo/pull/191

So is my above thesis correct? https://github.com/neos/neos-development-collection/pull/4674#issuecomment-1945752108

With this fixes we then finally can release another beta 😅

**Comment by mhsdesign** [comment]
Omitting the content stream was created event should be possible here: https://github.com/neos/neos-development-collection/blob/bc39546ae0b06344c5963eefff54d5b296bcb436/Neos.ContentRepository.Export/src/Processors/EventExportProcessor.php#L51

**Comment by mhsdesign** [comment]
Before i fiddle to much around with this pr and have to many reverts, i created a separate clean pr implementing what i wrote above: https://github.com/neos/neos-development-collection/pull/4914

File: Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php [hunk 3]
```diff
@@ -400,10 +400,8 @@ private function connectRestrictionRelationsFromParentNodeToNewlyCreatedNode(
         NodeAggregateId $newlyCreatedNodeAggregateId,
         DimensionSpacePointSet $dimensionSpacePointsInWhichNewlyCreatedNodeAggregateIsVisible
     ): void {
-        // TODO: still unsure why we need an "INSERT IGNORE" here;
```
Comment (mhsdesign) [inline Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:403]: this should probably be extracted into a separate change?

**Comment by ahaeslich** [comment]
Closed in favour of #4914

**Comment by mhsdesign** [comment]
Not so schnell ill reopen and restore the branch, as there are two changes in here that should be still made and extracted into separate prs.

Otherwise i would just have hacked this pr, but it just was to much.

- TODO: still unsure why we need an "INSERT IGNORE" here;
- and maybe also the removal of the Disable the same node twice tests.

... but these were not _my_ changes so i cant really create a pr and defend (understand) them ;)

**Comment by ahaeslich** [comment]
Then @nezaniel should probably be made aware of this ;). Because I asked him before closing it if this is still relevant.

**Comment by nezaniel** [comment]
I'd say we close this anyway; the tests test a different behavior than what is now implemented and the INSERT IGNORE is completely unrelated

**Comment by mhsdesign** [comment]
Ha funny and the obsolete `INSERT IGNORE` were fixed by some other pr. So fine with me ;)
... i wasnt able to cherry pick it, and there were no references left ...
